File: Mesh/GameViewController.m
#import "GameViewController.h"
#import "IcosahedronModel.h"
#import "Mesh-Swift.h"
@import OpenGLES;
@import SpriteKit;

@interface GameViewController ()

@property (weak, nonatomic) IBOutlet SKView *infoView;
@property (nonatomic) GameScene *gameScene;
@property (nonatomic) EAGLContext *context;
@property (nonatomic) GLKBaseEffect *effect;
@property (nonatomic) IcosahedronModel *icosahedronModel;
@property (nonatomic) IcosahedronVertex *prevVertex;
@property (nonatomic) IcosahedronVertex *currentVertex;
@property (nonatomic) GLKQuaternion prevQuaternion;
@property (nonatomic) GLKQuaternion currentQuaternion;
@property (nonatomic) float animationProgress;

@end

@implementation GameViewController
{
    GLuint _program;
    GLuint _vertexBufferID;
    GLKMatrix4 _modelViewMatrix;
}

- (void)dealloc
{
    [self tearDownGL];
}

- (void)viewDidLoad
{
    [super viewDidLoad];

    self.context = [[EAGLContext alloc] initWithAPI:kEAGLRenderingAPIOpenGLES2];

    GLKView *glkView = (GLKView *)self.view;
    glkView.context = self.context;
    glkView.drawableDepthFormat = GLKViewDrawableDepthFormat24;

    [self setUpGL];

    self.icosahedronModel = [IcosahedronModel new];

    self.currentVertex = self.icosahedronModel.vertexDict[@"C"];
    self.animationProgress = 1.0;
    self.prevQuaternion = GLKQuaternionIdentity;
    self.currentQuaternion = GLKQuaternionIdentity;

    [self setUpInfoView];

    [self.gameScene updateInfo:self.currentVertex];
}

- (void)setUpGL
{
    [EAGLContext setCurrentContext:self.context];

    self.effect = [GLKBaseEffect new];
    self.effect.light0.enabled = GL_TRUE;
    self.effect.light0.diffuseColor = GLKVector4Make(1.0, 1.0, 1.0, 1.0);
    self.effect.colorMaterialEnabled = GL_TRUE;

    float aspect = fabs(self.view.bounds.size.width / self.view.bounds.size.height);
    float width = 1.0;
    float height = width / aspect;
    self.effect.transform.projectionMatrix = GLKMatrix4MakeOrtho(-width / 2, width / 2, -height / 2, height / 2, 0.1, 100);

    glEnable(GL_DEPTH_TEST);
}

- (void)tearDownGL
{
    [EAGLContext setCurrentContext:self.context];

    glDeleteBuffers(1, &_vertexBufferID);

    glDeleteProgram(_program);
    _program = 0;

    self.context = nil;
    [EAGLContext setCurrentContext:nil];
}

- (void)setUpInfoView
{
    self.gameScene = [GameScene sceneWithSize:self.view.bounds.size];
    [self.infoView presentScene:self.gameScene];
}

- (void)update
{
    GLKMatrix4 baseModelViewMatrix = GLKMatrix4MakeTranslation(0.0, 0.0, -5.0);

    if (self.animationProgress < 1.0) {
        self.animationProgress += self.timeSinceLastUpdate;
        self.animationProgress = MIN(1.0, self.animationProgress);
    }

    GLKQuaternion modelQuaternion = GLKQuaternionSlerp(self.prevQuaternion, self.currentQuaternion, self.animationProgress);

    _modelViewMatrix = GLKMatrix4MakeTranslation(0.0, 0.0, 0.0);
    _modelViewMatrix = GLKMatrix4Rotate(_modelViewMatrix, GLKMathDegreesToRadians(180), 0.0, 0.0, 1.0);
    _modelViewMatrix = GLKMatrix4Rotate(_modelViewMatrix, GLKMathDegreesToRadians(150), 1.0, 0.0, 0.0);
    _modelViewMatrix = GLKMatrix4Multiply(_modelViewMatrix, GLKMatrix4MakeWithQuaternion(modelQuaternion));
    _modelViewMatrix = GLKMatrix4Multiply(baseModelViewMatrix, _modelViewMatrix);
}

- (GLKQuaternion)quaternionForRotateFrom:(GLKVector3)from to:(GLKVector3)to
{
    GLKVector3 normalizedFrom = GLKVector3Normalize(from);
    GLKVector3 normalizedTo = GLKVector3Normalize(to);

    float cosTheta = GLKVector3DotProduct(normalizedFrom, normalizedTo);
    GLKVector3 rotationAxis = GLKVector3CrossProduct(normalizedFrom, normalizedTo);

    float s = sqrtf((1 + cosTheta) * 2);
    float inverse = 1 / s;

    return GLKQuaternionMakeWithVector3(GLKVector3MultiplyScalar(rotationAxis, inverse), s * 0.5);
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
{
    UITouch *touch = [touches anyObject];
    CGPoint location = [touch locationInView:self.view];
    location.x -= CGRectGetMidX(self.view.bounds);
    location.y -= CGRectGetMidY(self.view.bounds);

    float radian = atan2f(-location.y, location.x);
    radian = radian > 0 ? radian : radian + 2 * M_PI;

    [self moveNextVertex:radian];
}

- (void)moveNextVertex:(float)theta
{
    self.prevVertex = self.currentVertex;

    float unit = 2 * M_PI / 5;
    if (0 < theta && theta <= unit) {
        self.currentVertex = self.prevVertex.head;
    } else if (unit < theta && theta <= unit * 2) {
        self.currentVertex = self.prevVertex.leftHand;
    } else if (unit * 2 < theta && theta <= unit * 3) {
        self.currentVertex = self.prevVertex.leftFoot;
    } else if (unit * 3 < theta && theta <= unit * 4) {
        self.currentVertex = self.prevVertex.rightFoot;
    } else {
        self.currentVertex = self.prevVertex.rightHand;
    }
    self.animationProgress = 0.0;

    self.prevQuaternion = self.currentQuaternion;
    GLKQuaternion relativeQuaternion = [self quaternionForRotateFrom:self.currentVertex.coordinate to:self.prevVertex.coordinate];
    self.currentQuaternion = GLKQuaternionMultiply(self.currentQuaternion, relativeQuaternion);

    [self.gameScene updateInfo:self.currentVertex];
}

#pragma mark - GLKViewDelegate methods

- (void)glkView:(GLKView *)view drawInRect:(CGRect)rect
{
    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
    glClearColor(0.0, 0.0, 0.0, 1.0);

    self.effect.transform.modelviewMatrix = _modelViewMatrix;
    [self.effect prepareToDraw];

    glGenBuffers(1, &_vertexBufferID);
    glBindBuffer(GL_ARRAY_BUFFER, _vertexBufferID);
    glBufferData(GL_ARRAY_BUFFER, sizeof(Vertex) * IcosahedronModelNumberOfFaceVertices, self.icosahedronModel.vertices, GL_STATIC_DRAW);

    glEnableVertexAttribArray(GLKVertexAttribPosition);
    glEnableVertexAttribArray(GLKVertexAttribNormal);
    glEnableVertexAttribArray(GLKVertexAttribColor);

    glVertexAttribPointer(GLKVertexAttribPosition, 3, GL_FLOAT, GL_FALSE, sizeof(Vertex), (GLvoid *)offsetof(Vertex, position));
    glVertexAttribPointer(GLKVertexAttribNormal, 3, GL_FLOAT, GL_FALSE, sizeof(Vertex), (GLvoid *)offsetof(Vertex, normal));
    glVertexAttribPointer(GLKVertexAttribColor, 4, GL_FLOAT, GL_FALSE, sizeof(Vertex), (GLvoid *)offsetof(Vertex, color));

    glDrawArrays(GL_TRIANGLES, 0, IcosahedronModelNumberOfFaceVertices);
}

@end

File: Mesh/GameViewController.swift
import GLKit
import OpenGLES

func BUFFER_OFFSET(i: Int) -> UnsafePointer<Void> {
    let p: UnsafePointer<Void> = nil
    return p.advancedBy(i)
}

protocol GLfloatValueContainer {
    var values: [GLfloat] { get }
}

struct Vertex: GLfloatValueContainer {
    let position: GLKVector3
    let color: GLKVector4

    static var size: GLint {
        return GLint(sizeof(GLKVector3) + sizeof(GLKVector4))
    }

    var values: [GLfloat] {
        return [
            position.x, position.y, position.z,
            color.r, color.g, color.b, color.a,
        ]
    }
}

extension CollectionType where Generator.Element: GLfloatValueContainer {
    func GLfloatValues() -> [GLfloat] {
        return self.flatMap { $0.values }
    }
}

let red = GLKVector4Make(1.0, 0.0, 0.0, 1.0)
let blue = GLKVector4Make(0.0, 0.0, 1.0, 1.0)

var vertices: [Vertex] = [
    Vertex(position: GLKVector3Make(-0.5, -0.5, 0.0), color: red),
    Vertex(position: GLKVector3Make(-0.5,  0.5, 0.0), color: red),
    Vertex(position: GLKVector3Make( 0.5, -0.5, 0.0), color: red),
    Vertex(position: GLKVector3Make( 0.5, -0.5, 0.0), color: blue),
    Vertex(position: GLKVector3Make( 0.5,  0.5, 0.0), color: blue),
    Vertex(position: GLKVector3Make(-0.5,  0.5, 0.0), color: blue),
]

var indices: [GLubyte] = [
    0, 1, 2,
    3, 4, 5,
]

class GameViewController: GLKViewController {
    var vertexBufferID: GLuint = 0
    var indexBufferID: GLuint = 0
    var context: EAGLContext!
    var effect: GLKBaseEffect!

    deinit {
        EAGLContext.setCurrentContext(self.context)

        if (vertexBufferID != 0) {
            glDeleteBuffers(1, &vertexBufferID)
        }

        if (indexBufferID != 0) {
            glDeleteBuffers(1, &indexBufferID)
        }

        self.context = nil
        EAGLContext.setCurrentContext(nil)
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        self.context = EAGLContext(API: .OpenGLES2)

        let view = self.view as! GLKView
        view.context = self.context

        self.setUpGL()
    }

    func setUpGL() {
        EAGLContext.setCurrentContext(self.context)

        let effect = GLKBaseEffect()
        effect.useConstantColor = GLboolean(GL_TRUE)
        effect.constantColor = GLKVector4Make(1.0, 1.0, 1.0, 1.0)
        self.effect = effect

        glClearColor(0.0, 0.0, 0.0, 1.0)
    }

    // MARK: - GLKViewDelegate methods

    override func glkView(view: GLKView, drawInRect rect: CGRect) {
        self.effect.prepareToDraw()

        glClear(GLbitfield(GL_COLOR_BUFFER_BIT))

        var triangles = vertices.GLfloatValues()

        glGenBuffers(1, &vertexBufferID)
        glBindBuffer(GLenum(GL_ARRAY_BUFFER), vertexBufferID)
        glBufferData(GLenum(GL_ARRAY_BUFFER), GLsizeiptr(sizeof(GLfloat) * triangles.count), &triangles, GLenum(GL_STATIC_DRAW))

        glGenBuffers(1, &indexBufferID)
        glBindBuffer(GLenum(GL_ELEMENT_ARRAY_BUFFER), indexBufferID)
        glBufferData(GLenum(GL_ELEMENT_ARRAY_BUFFER), GLsizeiptr(sizeof(GLubyte) * indices.count), &indices, GLenum(GL_STATIC_DRAW))

        let positionAttrib = GLuint(GLKVertexAttrib.Position.rawValue)
        let colorAttrib = GLuint(GLKVertexAttrib.Color.rawValue)

        glEnableVertexAttribArray(positionAttrib)
        glEnableVertexAttribArray(colorAttrib)

        glVertexAttribPointer(positionAttrib, 3, GLenum(GL_FLOAT), GLboolean(GL_FALSE), GLsizei(Vertex.size), BUFFER_OFFSET(0))
        glVertexAttribPointer(colorAttrib, 4, GLenum(GL_FLOAT), GLboolean(GL_FALSE), GLsizei(Vertex.size), BUFFER_OFFSET(sizeof(GLKVector3)))

        glDrawElements(GLenum(GL_TRIANGLES), GLsizei(indices.count), GLenum(GL_UNSIGNED_BYTE), nil)
    }
}
